dandi/upload.py

```diff
@@ -383,3 +402,46 @@ def check_replace_asset(
 
 def skip_file(msg: Any) -> Dict[str, str]:
     return {"status": "skipped", "message": str(msg)}
+
+
+def _bids_discover_and_validate(dandiset_path, paths, validation):
+    """Temporary implementation for discovery and validation of BIDS datasets
+
+    References:
+    - unification of validation records: https://github.com/dandi/dandi-cli/issues/943
+    - validation "design doc": https://github.com/dandi/dandi-cli/pull/663
+    """
+    from .utils import find_files
+    from .validate import validate_bids
+
+    if paths:
+        bids_lookup_paths = set(p for p in paths)
+    else:
+        bids_lookup_paths = None
+    bids_descriptions = map(Path, find_files("dataset_description.json", dandiset_path))
+    bids_datasets = [p.parent for p in bids_descriptions]
+    if bids_datasets:
+        lgr.debug(
+            "Detected %d BIDS datasets at following paths: %s",
+            len(bids_datasets),
+            ", ".join([str(i) for i in bids_datasets]),
+        )
+
+    if validation != "skip":
+        if bids_lookup_paths:
+            bids_datasets_to_validate = set()
```

Comment:
Change this to a list so that the type is the same regardless of `bids_lookup_paths`.